### ISSUE TITLE
predicted_peak_bytes under-predicts VRAM for an untied streamed head once #1061 lands; is #975's one-slot ping-pong worth trading for 2 slots?

### ISSUE THREAD

**Issue by VanshSharmaPES** [MakazhanAlpamys/Soup#1174]
Follow-up from #975, split into two independently-verifiable pieces per that thread's discussion.

## Part 1 — `estimate_stream_peak_vram` under-predicts for any untied streamed head

`estimate_stream_peak_vram` (`src/soup_cli/utils/layer_stream.py:1190`) documents its own contract: "worst absolute error 0.85%, and it never under-predicts — the only safe direction for a number that is allowed to refuse a run." Its `large_layer_bytes` parameter is deliberately "one reusable slot sized to the larger of `embed_tokens` and an untied `lm_head`; the two matrices no longer add together at peak (#324)" (`layer_stream.py:1216`).

PR #1061 (open, `fix/1049-untied-streamed-preference`) adds a second, temporary head-sized allocation on top of that slot: `StreamedLargeLayer.forward` now clones its weight (new `_needs_a_private_weight()` / the `weight = weight.clone()` line in `layer_stream_runtime.py`) whenever `torch.is_grad_enabled()` and the pool's large-layer slot is shared by more than one key and the module isn't an embedding — i.e. whenever the untied `lm_head` runs forward during training. That check isn't gated on loss type: the changelog frames it as a preference-loss (dpo/kto) fix (the bug it fixes only manifests there, since only preference losses run two forwards per step), but the guard itself is `torch.is_grad_enabled()` plus "slot shared, not embedding" — so it also fires on every SFT step against an untied streamed checkpoint, not just dpo/kto.

`_stream_budget_lines` (`src/soup_cli/trainer/stream_setup.py:887-896`) still calls `estimate_stream_peak_vram` with a single `large_layer_bytes=large_buffer_bytes` (`stream_setup.py:893`) and has no awareness of this new clone. Once #1061 merges, any untied streamed training run is under-predicted by roughly one head-sized buffer — for a Llama-3.1-8B-shaped untied head that's `128256 x 4096 x 2 bytes ~= 1.05 GiB`. Per the estimator's own stated contract, this needs fixing before or alongside #1061, not after someone hits the gap as a silent WDDM spill or an OOM the pre-flight was supposed to catch.

This part doesn't need special hardware — it's arithmetic against a documented condition, verifiable with a unit test against `_needs_a_private_weight`'s actual trigger (untied + grad-enabled), not against loss type. I can send a PR for this.

## Part 2 — is the #975 ping-pong stall worth trading VRAM to remove?

From #975: `LargeLayerBufferPool`'s single shared slot forces a full head-sized H2D copy every step (~33 ms on that issue's synthetic-70B/RTX 5070 fixture — a genuinely un-hideable pinned-memory transfer at that matrix size, not a `_plan_queue` disk-scheduling effect). Giving `embed_tokens` and an untied `lm_head` permanent, separate slots would remove that stall entirely, at the cost of one extra head-sized buffer resident for the whole run (same order of magnitude as Part 1's number).

Candidate directions to measure, in increasing order of design cost:
- **(a) current** — one shared slot, full-matrix H2D copy every step (measured in #975).
- **(b) reorder the prefetch** — issue `embed_tokens`'s `load_async` at the tail of the *backward* pass instead of at `_prime()` (the forward-pre-hook). Same VRAM cost as (a), but gives the copy real time to overlap with the tail of backward compute instead of firing with nothing to hide behind.
- **(c) two permanent slots** — removes the stall entirely, costs ~1 GiB extra VRAM held for the whole run, which matters specifically on the 8 GB cards layer streaming exists for.

The `wait()` bracket instrumentation in `benchmarks/harness/layer0_wait.py` / `stream_probe.py` is already the right tool to compare these — the embedding's stall should go to ~0 under (b) and (c). Deciding between (b) and (c) is a VRAM-vs-complexity tradeoff that needs measuring on real hardware (an RTX 5070 or similar), not reasoning from first principles.

I don't have that hardware here (a 4 GB RTX 3050 locally can't load the synthetic-70B fixture), so I can scope and build the harness variant for (b)/(c) but can't produce the comparison numbers myself.

## Links
- #975
- #1061
- #1049

**Comment by MakazhanAlpamys** [comment]
Thanks, @VanshSharmaPES — splitting it this way is right, and Part 1 is correct. I checked it against #1061's current head (`b3f54036`) rather than taking it on the reasoning:

```python
if torch.is_grad_enabled() and self._needs_a_private_weight():
    weight = weight.clone()
```

and `_needs_a_private_weight()` is "the slot is shared and this is not an embedding" — nothing about the loss. So an SFT step on an untied streamed checkpoint pays the head-sized copy too, and it is alive from the head's forward to its backward, which is exactly where the step peaks.

**Part 1 belongs in #1061, not in a separate PR.** Its estimator half is already a blocking item there (B2 in my last review: the private copy falsifies `predicted_peak_bytes`'s "never under-predicts" docstring, ~1.05 GiB on an 8B-class untied head). What you add is the more important half: SFT has one forward per step, nothing refills the shared slot between the head's forward and its backward, so SFT never needed the copy at all — it would pay ~1 GiB of peak for nothing, on the 8 GB cards streaming exists for. The clone and its accounting have to land together, so I am carrying this to #1061 and crediting you there; please don't open a second PR for it unless @dchaudhari7177 asks to hand it over.

**Part 2 is the real open question, and (b) has to be measured before (c).** (b) costs no VRAM; (c) holds ~1 GiB for the whole run on the cards that can least afford it, so it needs (b) shown insufficient first, with the `wait()` bracket numbers from `layer0_wait.py` / `stream_probe.py` for all three arms on the same card in the same session. If you want to take that measurement, say so here and name the card.

**Comment by VanshSharmaPES** [comment]
Understood on both counts.

**Part 1:** not opening a second PR for it — good to know it's already B2 in your #1061 review, and that the SFT-pays-for-nothing half is the more important addition. Thanks for the credit; happy for it to land there.

**Part 2:** I implemented (b) and opened it as a draft: #1178. Correctness is verified — full streaming suite unchanged (486 tests), a new targeted test confirms the backward-tail prefetch fires and `_prime()`'s own load becomes a same-owner no-op from the second step on, and 2 training steps complete with no autograd error. I also manually confirmed the DPO/KTO version error reproduces identically with and without this change, so it's #1049, not something (b) introduces.

I can't take the `wait()`-bracket measurement myself — my hardware is a 4 GB RTX 3050, same constraint as #975, and it can't load the cold-70B fixture. Opening it as a draft was meant to give whoever can run it something concrete to check out rather than a description. If you or someone else can get the (a)/(b)/(c) numbers on the 5070 (or similar), that's the piece I'm missing to close this out — otherwise I don't have a path to produce it.

**Comment by MakazhanAlpamys** [comment]
**Part 1 is closed by #1061, now merged.** The private head copy is taken only by `dpo` and `kto`, the losses whose reference forward runs before the backward. `sft`, `orpo` and `simpo` keep the view, so an SFT step on an untied streamed checkpoint no longer pays for a head-sized copy it never needed. The VRAM pre-flight charges the copy as a second large-layer slot for exactly the runs that take it, so `predicted_peak_bytes` stays on the safe side of its docstring. Thank you, @VanshSharmaPES: the SFT half was yours, and it changed what #1061 does, not only how it is accounted. #1061's changelog entry credits you for that scoping.

**Part 2 stays open here and continues on #1178.** Arms (a) and (b) are measured there: the embedding's `wait()` drops from 18.6-22.7 ms to 0.12-0.19 ms per step on a smaller untied fixture. With (b) in, what a second slot would still buy is the head's forward-tail wait (10.6-14.6 ms per step on that fixture). Moving the head's load earlier in the forward costs no VRAM, so that gets measured before (c).

**Comment by VanshSharmaPES** [comment]
Thanks for running (a) and (b) on the 5070. The results are in #1178, and I've pushed the fixes from your two reviews there.

**Where that leaves Part 2.** With (b) in, the embedding's wait goes from about 22 ms to about 0.15 ms on your fixture, so what's left of the large-layer wait is the head's forward-tail load, 10.6-14.6 ms per step. Part 1 stays with #1061 as you said, and I'm not opening anything for it.

**I agree (c) shouldn't come next.** Your point that the embedding's bytes in the slot are dead once the lookup has run looks right to me from the code, and it changes what (c) is competing against. Today `tail_prefetch` fires from `StreamPrefetcher.advance()` for the last decoder layer, before that layer's `functional_call` (`_body` calls `pool.wait`, then `advance`, then `functional_call`). So the head's copy can only hide behind one decoder layer's forward. Issuing it right after the embedding lookup instead would give it every decoder layer's forward to hide behind, for no extra VRAM. On your 8-layer fixture that's roughly 8x the overlap window. That's a much cheaper thing to try before spending a GiB on a second slot.

**Hazards I'd expect it to have to clear**, from reading the code and not from running anything:
- The embedding lookup kernel is queued on the compute stream and reads the slot, so the refill has to be ordered after it. `wait_stream` in `LargeLayerBufferPool.load_async` should cover that the same way it covers the head's backward in #1178, but it would want its own pool-level GPU test, like the one I just added.
- The head must survive until its own forward and, if #1061's clone applies, until it has been cloned. Nothing else refills the slot in between, but that assumption needs a test.
- Eval and no-grad forwards take the same path, so they'd need checking too.

**Suggested next step.** If you'd like it, I'll prototype that on a separate branch and check correctness on CPU the way I did for (b): bit-equality against `main`, the load counts, and the mutation checks. I can't measure the wait myself, since I only have a 4 GB RTX 3050 and my torch is currently a CPU build, so the number would have to come from your 5070 again. I'd rather have your go-ahead before opening a second PR while #1178 is still in draft, and I'd hold (c) until that number is in.

I haven't taken the two other non-blocking items from your review, the copy-stream sync in `StreamRuntime.close()` and the redundant `backward_tail_prefetched` flag, to keep #1178 narrow.

**Comment by MakazhanAlpamys** [comment]
@VanshSharmaPES, go ahead. Open the prototype as a separate draft PR now; you don't need to wait for #1178 to leave draft. Holding (c) until there is a number for this is the right order.

Your reading of today's timing is right. `tail_prefetch` fires from `StreamPrefetcher.advance()` at the last decoder layer (`layer_stream_runtime.py:1017-1024` on `main`). `_body` calls `pool.wait`, then `advance`, then `functional_call`, so the head's copy overlaps only that one layer's forward. The three hazards you list are the right ones. Reading the code for your proposal turned up two more. Neither is measured, but both change what the prototype has to show.

**1. The copy stream, not the layer count, decides how much the head can hide behind.** `install_streaming` creates one copy stream (`layer_stream_runtime.py:2607`), and both pools use it. `advance()` prefetches only `idx + direction` (`:1014`). A head copy issued right after the lookup therefore sits in that in-order stream ahead of the next decoder copy the prefetcher issues, and every later decoder copy waits behind it.

At a step start, the backward has just left layers 0 and 1 resident. I checked this on CPU: over 3 steps of a 4-layer untied fixture, no step after the first reloads layer 1 at its start. So with the default `stream_buffers: 2`, the first copy queued behind the head is layer 2's, and the head can hide behind the compute of layers 0 and 1, not of all 8. Whatever it does not cover turns into a wait on a decoder entry instead of on the head.

That can still beat today's one-layer window, but it is not 8x. If it falls short, the options are a stream of its own for the head (it then shares PCIe bandwidth with the decoder copies instead of queueing ahead of them), or issuing it at a later layer. Keep this in mind when you choose where the call goes.

**2. On the disk tier, the head's `get()` would come at a point in the decoder walk where no vocabulary `get` comes today.** `AsyncDiskSource.get` replaces the reader's lookahead queue with a fresh plan on every hit (`async_disk_source.py:1022`). The docstring of `_note_direction` (`:945`) lists where a vocabulary `get` can come today: before the decoder walk, at the forward turnaround, and at the step wrap. A head `get` just after the lookup comes in the middle of the forward walk. Its hit plans from the head's own single-member group, which yields nothing, so a decoder read that was planned but not yet started can be dropped and become a demand read later.

Whether that happens depends on timing, so I can't say it will. It does mean the disk tier has to be measured, not inferred. And if the prototype needs any change to `_plan_queue`, `_claim_slot` or `get`, #975's race-battery constraint applies again. #1178 does not have this problem. On the same 4-layer fixture over 3 steps, its `source.get()` sequence is identical to `main`'s, plus one embedding `get` at the end of `train()`.

**For the CPU half, bring the same evidence as for (b):**

- a switch that restores today's behaviour (a callback left as `None`, the way `backward_tail_prefetch=None` gave arm A in #1178), so both arms can run interleaved in one process;
- bit-equality with #1178 on CPU fp32 for untied SFT (also at `gradient_accumulation_steps: 2`), ORPO, SimPO, DPO and KTO (#1061 is on `main` now), and tied SFT, where the switch should do nothing;
- the large-layer load count per step, including a no-grad or eval forward followed by a training step;
- mutations of the main change, with counts both ways;
- a pool-level GPU test for the new refill point, where the queued read is the embedding lookup instead of the head's backward, skipped without CUDA like the one in #1178.

Branch it from #1178's head and say so in the description. The arm that matters is (b) against (b) plus the moved head load, and the branch will need a rebase once #1178 lands.

**The measurement.** Run it on #1178's smaller fixture, not the cold-70B one. With (b) in, the head's wait there is 10.6-14.6 ms per step. On #975's cold-70B record the `lm_head` entry is 0.002 ms per step, because the last decoder layer's load already hides it, so that fixture would show nothing either way. Judge the result by the step's total wait, every entry in `stalls`, on both the RAM and the disk tier. Point 1 is why: the head's own entry can improve while the step does not.

The 5070 is busy today; a measurement will be scheduled on it when it is free. The CPU work doesn't need to wait for that.

**One more zero-VRAM candidate, found while checking this and not measured:** on `main`, `StreamPrefetcher.prime()` re-copies layer 0 at every step start even when it is already resident (`layer_stream_runtime.py:1002`, unchanged since `00833ac7`). Its cost is unknown. If your measurement shows a layer-0 wait at the step start, this is the likely source; fold it into the prototype or leave it for its own issue, your call.
